### DISCUSSION TITLE
aioble examples for multiple peripherals

### DISCUSSION THREAD

**Discussion by simonwgill** [micropython/micropython#12674] [Using MicroPython]
I've been having some weird issues trying to connect to multiple devices at once. I expect that it should work, but I'm not all that comfortable with the way I'm doing the connections.

It would be really helpful to have a version of the temp_client example that can connect and read from as many temperature sensors are nearby to show the idiomatic way to do device management. If there's any factors that would affect the number of simultaneous connections, it would be very helpful to have a pointer to those as well.

Thanks,
Simon

**Comment by brianreinhold** [comment]
You have to at least configure the bluetooth stack you are using to support more than one device. Then you have to allow the scanner to restart after connecting as the scanner is stopped during connection. I do this by connecting, doing some basic things that all devices need like service discovery, reading the DIS, CTS, and Battery service, and then creating tasks to handle the device-specific stuff. After the task is created I exit the connection method allowing the scanner to restart.

**Reply by simonwgill** [reply]
Thanks for the response, I wasn't sure I would ever get one.

I can't see any way to modify the number of possible connections in aioble. There does seem to be a [value in pico-sdk](https://github.com/raspberrypi/pico-examples/blob/eca13acf57916a0bd5961028314006983894fc84/pico_w/bt/standalone/btstack_config.h#L26) but I think that would involve building a fork of micropython to open up.

If it matters, I'm using a Pico W for this, not sure how it would differ across the different controllers and bluetooth chips.

**Reply by brianreinhold** [reply]
I too am using the PICO_W. You cannot configure the number of simultaneous connections in aioble as each stack needs its own way of setting up memory for such endeavors and thus it is port dependent. When you build the Micropython firmware for the PICO port (btstack) there will be #defines where you set the number of connections.

I changed these two
#define MAX_NR_GATT_CLIENTS 1
#define MAX_NR_HCI_CONNECTIONS 1
to
#define MAX_NR_GATT_CLIENTS 3
#define MAX_NR_HCI_CONNECTIONS 3
I am not sure I needed to change the number of GATT clients; I am not sure what that parameter means.
THese are located in /extmod/btstack/btstack_config_common.h in the MP distribution.

**Reply by CraigGT** [reply]
I'm working on a project that needs two connections and thanks to @brianreinhold's reply I've got the connections established.
Of course I had to learn how to rebuild the firmware first but that's another story 😁

**Comment by brianreinhold** [comment]
It was a challenge figuring out how to build MP.  I hope your communications don't need (Bluetooth) pairing. That's a whole new level of headache as pairing is not supported for the PICO_W in the MP distribution. You will have to do a significant amount of coding in modbluetooth_btstack.c and related files.

**Comment by CraigGT** [comment]
Hopefully not,
I already have a one device version fully working and the two device version can maintain two connections and retrieve the services and characteristics and retrieve battery levels so it's not far off.

I rebuilt MP for two connections, is there a big memory hit for adding more?

**Reply by brianreinhold** [reply]
I never did a check on the memory requirements so I can say. Anything I would say would be pure speculation.

I don't know what kind of devices you are working with but if you are able to (fully) work with them they obviously don't need pairing/bonding. On the other hand, if these are health devices you still need to get measurements. Health devices that require pairing may send security requests or give you an insufficient authentication error when you try to read/write some characteristic or descriptor.

**Reply by CraigGT** [reply]
I saw somewhere that each connection needed its own memory stack so I just changed the defines you suggested to 2 because that's how many I need but the OP said he required 'many'

My devices are more 'adult' than health so only need regular data sends but experience with different versions has shown me that getting the connections and characteristics is the hard part, sending the data is the easy part.

Yours is the only answer I've seen that explains why I was failing trying to make the 2nd connection so thanks again.
